### PR TITLE
[sailfish-browser] Ability to choose start page as default instead of URL. Fixes JB#50678

### PR DESCRIPTION
**Pull request by aroditelev**

### CODE REVIEW

File: apps/browser/qml/pages/SettingsPage.qml [hunk 1]
```diff
@@ -40,21 +46,32 @@ Page {
...
+                    MenuItem {
+                        //% "Default home page"
+                        text: qsTrId("sailfish_browser-me-home_page_default")
+                        onClicked: homePageConfig.value = "about:blank"
+                    }
+                    MenuItem {
+                        readonly property string site: removeProtocolTypeFromUri(homePageConfig.value)
+                        //% "Custom website %1"
+                        property string title: homePageConfig.value === "about:blank" ? qsTrId("sailfish_browser-me-home_page_custom").arg("")
+                                                                                      : qsTrId("sailfish_browser-me-home_page_custom").arg(site)
```
Comment (xlise) [inline apps/browser/qml/pages/SettingsPage.qml:68]:
```suggestion
                        property string title: qsTrId("sailfish_browser-me-home_page_custom").arg(homePageConfig.value === "about:blank" ? "" : site)
```

Comment (aroditelev) [reply apps/browser/qml/pages/SettingsPage.qml]: Yes it looks better thanks)

File: apps/browser/qml/pages/components/AddHomePageDialog.qml [hunk 2]
```diff
@@ -0,0 +1,63 @@
...
+import QtQuick 2.0
+import Sailfish.Silica 1.0
+import Sailfish.Policy 1.0
+import org.nemomobile.configuration 1.0
+
+Dialog {
+    id: root
+    property alias text: textField.text
+    property alias placeholderText: textField.placeholderText
+    property alias description: textField.description
```
Comment (llewelld) [inline apps/browser/qml/pages/components/AddHomePageDialog.qml:20]: The three aliases and id look unused, so can be removed.

File: apps/browser/qml/pages/components/AddHomePageDialog.qml [hunk 3]
```diff
@@ -0,0 +1,63 @@
...
+
+    function containsSpaces(text) {
+       return /^\s+|\s+$|\s*(\s)/.test(text) && text.length > 0
+    }
+
+    canAccept: (textField.text !== "" && !textField.errorHighlight)
```
Comment (llewelld) [inline apps/browser/qml/pages/components/AddHomePageDialog.qml:26]: As far as I can tell the `textField.text !== "" ` condition is already covered by `containsSpaces()` so can be removed from here.

File: apps/browser/qml/pages/components/AddHomePageDialog.qml [hunk 4]
```diff
@@ -0,0 +1,63 @@
...
+    onAccepted: {
+        homePageConfig.value = textField.text || "about:blank"
+    }
+
+    TextField {
+        id: textField
+
+        width: parent.width
+        anchors.top: header.bottom
+        enabled: AccessPolicy.browserEnabled
```
Comment (llewelld) [inline apps/browser/qml/pages/components/AddHomePageDialog.qml:44]: With the settings pages included as part of the browser, are these AccessPolicy restrictions still needed?

File: apps/browser/qml/pages/components/AddHomePageDialog.qml [hunk 5]
```diff
@@ -0,0 +1,63 @@
...
+    }
+
+    onAccepted: {
+        homePageConfig.value = textField.text || "about:blank"
+    }
+
+    TextField {
+        id: textField
+
+        width: parent.width
```
Comment (llewelld) [inline apps/browser/qml/pages/components/AddHomePageDialog.qml:42]: I believe `TextField` defaults to the parent width, so this can be removed.

File: apps/browser/qml/pages/SettingsPage.qml [hunk 6]
```diff
@@ -41,21 +47,32 @@ Page {
...
+                iconSource: "image://theme/icon-m-home"
+                menu: ContextMenu {
+                    MenuItem {
+                        //% "Default home page"
+                        text: qsTrId("sailfish_browser-me-home_page_default")
+                        onClicked: homePageConfig.value = "about:blank"
+                    }
+                    MenuItem {
+                        readonly property string site: removeProtocolTypeFromUri(homePageConfig.value)
+                        //% "Custom website %1"
```
Comment (llewelld) [inline apps/browser/qml/pages/SettingsPage.qml:67]: It might be helpful for the translators if you could explain what the `%1` will be replaced with.

File: apps/browser/qml/pages/components/AddHomePageDialog.qml [hunk 7]
```diff
@@ -0,0 +1,63 @@
...
+        text: homePageConfig.value === "about:blank" ? "" : homePageConfig.value
+        //% "Enter website address that should be shown when browser is opened with no tabs to load."
+        description: qsTrId("sailfish_browser-he-home_page_website-address")
+        //% "Website address"
+        placeholderText: qsTrId("settings_browser-la-website_address")
+        label: placeholderText
+        inputMethodHints: Qt.ImhNoPredictiveText | Qt.ImhNoAutoUppercase | Qt.ImhUrlCharactersOnly
+        EnterKey.iconSource: "image://theme/icon-m-enter-close"
+        EnterKey.onClicked: focus = false
+        errorHighlight: containsSpaces(text)
```
Comment (llewelld) [inline apps/browser/qml/pages/components/AddHomePageDialog.qml:54]: It's customary to switch out the description with some info about why there's an error in case the dialog can't be accepted. For example, something like "Spaces can't be included in the URL".

Is it necessary to exclude spaces from the URL?

File: apps/browser/qml/pages/SettingsPage.qml [hunk 8]
```diff
@@ -41,21 +47,32 @@ Page {
...
-
-                inputMethodHints: Qt.ImhNoPredictiveText | Qt.ImhNoAutoUppercase | Qt.ImhUrlCharactersOnly
-
-                onTextChanged: homePageConfig.value = text || "about:blank"
-
-                EnterKey.iconSource: "image://theme/icon-m-enter-close"
-                EnterKey.onClicked: focus = false
+                //% "Default"
+                value: homePageConfig.value === "about:blank" ? qsTrId("sailfish_browser-la-home_page_default") : removeProtocolTypeFromUri(homePageConfig.value)
+                iconSource: "image://theme/icon-m-home"
```
Comment (llewelld) [inline apps/browser/qml/pages/SettingsPage.qml:58]: There's an animation glitch when the item is changed from a long URL, to "Default home page". In this case the little "Home" icon jumps to the floor before moving up to the centre of the item. It must be an issue with `BrowserListItem`; would it be possible to fix it?

**Review by llewelld — CHANGES_REQUESTED** [review]
This is already looking really nice. I've added a few comments.

File: apps/browser/qml/pages/components/AddHomePageDialog.qml (hunk 3 above)
Comment (aroditelev) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml:26]: containsSpaces will be relevant when `text.length > 0`

Comment (llewelld) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml:26]: Yes, you're absolutely right, my mistake. Thanks for clarifying.

In this case, maybe the error highlight should show if the textField is empty, in order to clarify to the user why the dialogue can't be accepted. What do you think?

File: apps/browser/qml/pages/components/AddHomePageDialog.qml (hunk 2 above)
Comment (aroditelev) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml]: @llewelld  Done

File: apps/browser/qml/pages/components/AddHomePageDialog.qml (hunk 4 above)
Comment (aroditelev) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml]: @llewelld  Done

File: apps/browser/qml/pages/components/AddHomePageDialog.qml (hunk 5 above)
Comment (aroditelev) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml]: @llewelld  Done

File: apps/browser/qml/pages/components/AddHomePageDialog.qml (hunk 7 above)
Comment (aroditelev) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml:48]: @llewelld  Done.  Added by "Spaces can't be included in the URL", but there was no corresponding inscription in the design. 
`Is it necessary to exclude spaces from the URL?` Сombobox will have an extra space at the beginning of the line, and desktop browsers remove spaces in the address automatically.

File: apps/browser/qml/pages/SettingsPage.qml (hunk 8 above)
Comment (aroditelev) [reply apps/browser/qml/pages/SettingsPage.qml:58]: Most likely this is a problem in `BrowserListItem`

File: apps/browser/qml/pages/SettingsPage.qml (hunk 6 above)
Comment (aroditelev) [reply apps/browser/qml/pages/SettingsPage.qml:68]: @llewelld Done

File: apps/browser/qml/pages/components/AddHomePageDialog.qml (hunk 3 above)
Comment (aroditelev) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml:21]: @llewelld  `In this case, maybe the error highlight should show if the textField is empty, in order to clarify to the user why the dialogue can't be accepted. What do you think?` I added to the description: ` //% "Enter website address that should be shown when browser is opened with no tabs to load."
        description: !errorHighlight ? qsTrId("sailfish_browser-he-home_page_website-address")
                                       //% "Spaces can't be included in the URL."
                                     : qsTrId("sailfish_browser-he-home_page_website-spaces_error")`

File: apps/browser/qml/pages/SettingsPage.qml (hunk 8 above)
Comment (aroditelev) [reply apps/browser/qml/pages/SettingsPage.qml:58]: @llewelld  added an overload alias: `property alias animationDuration: valueButton._duration` in` BrowserListItem`. Now "Home" icon not jumps.

Comment (llewelld) [reply apps/browser/qml/pages/SettingsPage.qml:58]: Thanks for taking a look at this. I'm still getting the icon jumping position (but in a slightly different way). It's more obvious if text size is set to Huge.

I agree with you that this is a problem in `BrowserListItem`, so if you prefer we can leave it as it is here and look into the jump separately.

File: apps/browser/qml/pages/components/AddHomePageDialog.qml [hunk 9]
```diff
@@ -0,0 +1,57 @@
...
+
+    onAccepted: {
+        homePageConfig.value = textField.text || "about:blank"
+    }
+
+    TextField {
+        id: textField
+        anchors.top: header.bottom
+        text: homePageConfig.value === "about:blank" ? "" : homePageConfig.value
+        //% "Enter website address that should be shown when browser is opened with no tabs to load."
```
Comment (llewelld) [inline apps/browser/qml/pages/components/AddHomePageDialog.qml:38]: Thank you for adding this. I'd like to suggest some small changes to simplify the wording:
```suggestion
        //% "Shown when the browser is opened with no tabs to load"
```

File: apps/browser/qml/pages/components/AddHomePageDialog.qml (hunk 7 above)
Comment (llewelld) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml:48]: Could you use `trim()` to remove the extra spaces instead for the combo box instead (or maybe I'm misunderstanding)?

It's just that we allow spaces in the address bar, and in favourites, so it would seem reasonable to allow them here too (otherwise the user has to manually change each of them to a %20 instead).

**Review by llewelld — COMMENTED** [review]
Thanks for making the changes.

File: apps/browser/qml/pages/SettingsPage.qml (hunk 8 above)
Comment (aroditelev) [reply apps/browser/qml/pages/SettingsPage.qml]: @llewelld it seems to me better to abandon the `BrowserListItem` in favor of the ComboBox. The problem is being solved.

File: apps/browser/qml/pages/components/AddHomePageDialog.qml (hunk 7 above)
Comment (aroditelev) [reply apps/browser/qml/pages/components/AddHomePageDialog.qml:48]: @llewelld  regexp finds all spaces, if, for example, the user enters https: // goo gle.com, then when the browser starts, not the page `https: // google.com` will open, but a search query, in the field of which there will be` https: // goo gle .com`. It checks and warns the user that this should not be done.

**Review by llewelld — APPROVED** [review]
Thanks, this works nicely and looks good to me.

**Review by atatarov — APPROVED** [review]
